docs/data/date-pickers/custom-views/custom-views.md

```diff
@@ -0,0 +1,51 @@
+---
```

Comment:
Why this page?
I saw 4 places where these new examples could land:


1. In the *Date Calendar* page, like for the `day` slot.
  I don't like this solution because we end up with some customization examples that are in the "Visual customization pages" and other that are in one of the component pages. I'm in favor of moving the `day` slot doc next to the new `monthButton` and `yearButton` slot doc wherever it ends up being. In the Date Calendar page, we could keep the "Week picker" example which is not just visual customization but also behavior customization.

2. In a "Date visual customization" (or other name) in the "Date components" folder. This one is probably the worse for me, it's not coherent with the other examples which are in "Visual customization" AND it's not coherent with the message we say in "Date Picker" that the shared features are in "Date Calendar". But it's easier to find from the navbar. 

3. In the "Custom slots and subcomponents" page. This would make a lot of sense, but this page is getting bigger and bigger and I'm looking for a way to split it.

4. In a new "Custom views" or "Custom date views" inside "Visual customization". I went for this approach because it's inside "Visual customization" (coherent) and it's a big enough sub-set of the slots to justify a dedicated page. I'm not sure if the time views should be documented in the same page though.

My idea for the overall structure of "Visual customization" would be to create an "Overview" page that introduces the concept of slots (an enhanced version of the first sections if "Custom slots and subcomponents"). This may link to some doc at the X level to avoid duplication (I need to create the issues about our Slack discussion around this page).
The current "Custom slots and sub-components" would loose it's introduction (moved to "Overview"), would be renamed "Other slots and sub-components" and would move to the end of the folder.

We would end up with something like this:

- Visual customization
  - Overview
  - Custom layout
  - Custom views
  - Custom field
  - Custom opening button
  - Other slots and sub-components

Reply:
Very nice reflections ! IMHO The new section makes sense and it looks clean on the docs 👏 
It made me think we could get rid of the repetition of the therm "custom", since it's already mentioned on the parent section. Something like:

- Visual customization
    - Slots and subcomponents
    - Layout
    - Views
    - Field
    - Opening button
    - Playground

Reply:
We would have to make sure it doesn't impact the Algola search, probably by only removing the term "custom" in the navbar, but not in the page title.
But I do agree that it would be visually lighter.

Reply:
@LukasTy @joserodolfofreitas I'd be interested by your opinion on this one

Reply:
**Regarding the nav bar naming**: given that this is not a menu -> submenu, but just a sub-section of the navbar, I'd stick with the existing naming, especially since the `VIsual customization` is just a styling element that would have no role in the accessibility—it's not an interactive element and it doesn't have any relationship with the underlying subsection items.

**Regarding the page structure: I'm not against a better organization, but some of the information in our docs is too scattered throughout. 🙈 
Let's take this proposal: why the `calendarHeader` slot deserves a spot on this page, but `arrowSwitcher` slots remain on the existing page? 🤔 
WDYT if we would use the following logic to define where a section should be put:
- Does it affect only a particular component? Place it on that component page.
- Does it affect multiple components (or/and view renderers)? Place it on the custom slots and subcomponents.

Reevaluate when such pages feel REALLY "too big".
A [React Spectrum—DateField](https://react-spectrum.adobe.com/react-spectrum/DateField.html) page for reference. I don't find the page too long. Maybe props could be separated, but other than that it seems fine. Colocation is also great for easier information discoverability. 🤷

Reply:
> Let's take this proposal: why the calendarHeader slot deserves a spot on this page, but arrowSwitcher slots remain on the existing page? 🤔


Because `arrowSwitcher` is also used in `TimeClock` :laughing: 
But the more I think about it, the more I want to destroy this abstraction and document how to override those icons / buttons next to the calendar header as you suggest.

> Does it affect multiple components (or/and view renderers)? Place it on the custom slots and subcomponents.

So all the custom field / custom opening button / custom layout demos would move back to "custom slots and subcomponents"?


> Colocation is also great for easier information discoverability. 

But we would rarely colocate with the component the user uses.
Most of the time, the dev that wants to customize let say the `calendarHeader` is using `DatePicker`, `DateTimePicker` or `DateRangePicker`.
And for me it is unclear if discoverability is better for him/her with the demo in `DateCalendar` rather than in "Visual customization".

Reply:
> Because `arrowSwitcher` is also used in `TimeClock` 😆

Well, isn't it also a view? 🤔 Given that the page is called `Custom views`.


> So all the custom field / custom opening button / custom layout demos would move back to "custom slots and subcomponents"?

Not exactly, I was referring to the views customization mainly.
Fields and layout are somewhat different "paradigms". I meant that I'm 55% for not creating this page and adding `monthButton` and `yearButton` slots showcasing in `Custom slots and subcomponents` or `DateCalendar`. 😉 

---

I agree that given the current structure, proper linking (referencing) of related pages is necessary. 👍 
Regarding the future of the document page structure—we probably shouldn't be concerned with it for now, until even `material-ui` is not using it. 🤔

Reply:
> Well, isn't it also a view? 🤔 Given that the page is called Custom views.

You are right :grimacing: 

> I meant that I'm 55% for not creating this page and adding monthButton and yearButton slots showcasing in Custom slots and subcomponents or DateCalendar. 😉

For "Custom slots and subcomponents", my main problem is that from a user perspective, I don't think it make sense to have some slots in a page named like this, other slots in a dedicated page and other slots in a component page from a component they don't directly used.

With the pickers we need to put some elements in common to avoid duplication, I think we all agree with that, but I feel like we are creating too many different groups that makes the navigation hard.

From a dev perspective, IMHO the reasoning is the same when looking for how to create a custom field or how to create a custom header. They won't think that the field slots are also used for Time Pickers so it make sense not to have them at the component level whereas the header is only for Date Pickers so it should live in `DateCalendar`.
Which is why I think it's easier to just put everything is "Visual customization" and correctly link from the component page (and if the slots table could be automatically generated in the component page instead of those API pages nobody uses correctly, it would be even better).

Reply:
@LukasTy not sure how we can move forward with this topic and have a structure we don't re-evaluate in the coming year (whatever structure it ends up being of course).

Reply:
> not sure how we can move forward with this topic and have a structure we don't re-evaluate in the coming year (whatever structure it ends up being of course).

I think that we can't. 🙈 
There are too many options we can take on it.
Besides, cramming more features always brings this same topic up back again. 🤷 😆 
If you feel that your suggestion makes the most sense, let's go with it until we feel it no longer works—just like we used to do up until now. 👌

Reply:
> You're pessimistic

I'm just trying to rely on past experience and the fact that we still have plans for new features and missing documentation, hence, I think this question will pop-up in the not-too-distant future. 😄 

> If you could describe more in depth the structure you have in mind, maybe we can come to an agreement on some structure.

My thought process is:
- put documentation as close to the source component as possible—Date Calendar in this case
- put in `Custom slots and subcomponents` whatever impacts multiple components (i.e. `DateCalendar` and `DateRangeCalendar`)

But I'm not saying that it is an ideal solution. Take for example this problem: where would the `monthButton` and `yearButton` examples end up if we add them in the `DateRangeCalendar`? 🤔 

My main confusion in this case is the question: Should the `Custom views` page be created or if we should add more demos in the `Custom slots and subcomponents`...

Reply:
I'm in favor of taking some time to discuss this topic in depth now, because in Q3 I plan to add the doc for the missing slots, so the topic will clearly rise again if we don't have an agreement.
The solution will not be perfect, I think can all acknowledge that it's a question of trade-offs and no solution can be optimal for all components.

The goal of our doc structure is that someone coming to the doc finds the topic he is looking for as easily as possible.
And what bothers me in both your proposal and mine is that it's that the place the doc he's looking for lives is far from obvious.
Imagine someone that is looking on how to customize the header of calendar in his `DateTimePicker`, he's probably entering the doc through the `DateTimePicker` component page or API page, neither of them are containing the information. 
If he's in the API page it's a dead-end (but that's another topic)

If he's in the component page, and the slot is in `DateCalendar`, he needs to:
  -  see the "Component composition" section
  - understand that the header is part of the `DateCalendar` component and not something at a higher level like `layout`, which is far from obvious
  - go to the `DateCalendar` component page
  - find the slot documentation

If he's in the component page, and the slot is in `Visual customization` => `Custom views`, he needs to:
  - see the "Visual customization" group of pages in the navbar
  - understand that the header is part of the view and not something at a higher level like `layout`, which is far from obvious
  - go to the "Custom views" page
  - find the slot documentation (easier than for `DateCalendar` because this page only contains slots doc)

And I'm pretty sure that some people will try to go in the `DateCalendar` and some people will try to go in the `Visual customization`.

I would be in favor of adding in each component page a "Sub components (slots)" section that lists the slots (could be the same table as in the API page, I still think the split between the API page and the component demo page is an aberration) with for each slot a link to where it is document (wherever that is).
This reduces a lot the importance of where the section actually is and would solve most of the problem IMHO.

Reply:
Concerning the place where we put the demo.
My main problem with putting some slots  in `DateCalendar` is that we split the slots in 2 big categories (the one documented in component demo pages and the one documented in "Visual customization").
And when a user comes to the doc to learn how to customize his component (most likely a `DatePicker` or a `DateTimePicker` in this scenario), he doesn't care if his slot is available on other components like the `TimePicker`.
So having the `calendarHeader` or the `field` in some place of the doc and the `monthButton` in a totally different place is not great IMHO.

That's why I would prefer to have everything under "Visual customization", with a good work on grouping them in meaningful pages.
And we can totally decide to put more demos in "Custom slots and subcomponents" instead of creating new pages, with maybe an additional level of depth in the structure of this page to help the navigation.


_____


TLDR: My proposal is the following:

1. On each component demo page a table / list with all the slots of this component. For each slot, we have a link to the section the documents it.
2. All the slots are documented under "Visual customization" to have a single place for people to look for if they don't use 1.

Reply:
I apologize for the late reply and the confusion I might have caused. 🙈 

I feel that we both are talking about the same thing, just in different words.
The point I wanted to make is that I did not think that we need to introduce a new page for slots, at least until `Custom slots and subcomponents` seems to work. 🤔 
I might have overstressed the option of moving this demo to the `DateCalendar` instead. 🙈 

Would you agree with the following proposal for page selection?
- Put the demo in the relevant component page (i.e. DateCalendar) if it's a "first-party" prop that impacts the component itself
- Choose a `Visual customization` section page if it's a slot that could be used in multiple Picker components;

The `day` slot is a bit of an exception, but I think that we can keep it out of this PR context. 👌 


___

In regards to the user journey finding the relevant demos, I think we should not forget about these 2 points you didn't mention:
1. quite a lot of users might end up resorting to the search if they don't quickly find what they want—thus, we should always keep in mind the creation of relevant headers for sections to cater to the search;
2. having too many places for where to look for the customization might cause more confusion—that's why I'm leaning towards sticking with a single `Custom slots and subcomponents` page as long as it makes sense;

___

> TLDR: My proposal is the following:
> 
> 1. On each component demo page a table / list with all the slots of this component. For each slot, we have a link to the section the documents it.

Great idea! 👍 Just one concern: if it's not automated as the API<->Demos relation currently is, I fear that it could get stale too quickly. 🙈 

> 2. All the slots are documented under "Visual customization" to have a single place for people to look for if they don't use 1.

I completely agree, I'm just voting for sticking with the `Custom slots and subcomponents` page for these new slots as well, but I'm not against the introduction of this page as well, it's just my preference, but I leave the decision to you. 😉

Reply:
> I apologize for the late reply and the confusion I might have caused. 🙈

Don't worry! We all have lots of topics in parallel.

> quite a lot of users might end up resorting to the search if they don't quickly find what they want—thus, we should always keep in mind the creation of relevant headers for sections to cater to the search;

True!

> having too many places for where to look for the customization might cause more confusion—that's why I'm leaning towards sticking with a single Custom slots and subcomponents page as long as it makes sense;

And do you think the existing pages we have outside of "Custom slots and subcomponents" are all justified? Or would you like to merge some of them back into the main page?

> Great idea! 👍 Just one concern: if it's not automated as the API<->Demos relation currently is, I fear that it could get stale too quickly. 🙈

Clearly, it's automated or it's dead :grimacing: 

> I completely agree, I'm just voting for sticking with the Custom slots and subcomponents page for these new slots as well, but I'm not against the introduction of this page as well, it's just my preference, but I leave the decision to you.

We can stick with the single page for now and reconsider in the future :+1:

Reply:
> And do you think the existing pages we have outside of "Custom slots and subcomponents" are all justified? Or would you like to merge some of them back into the main page?

The biggest offender seems to be the `Custom opening button`. It's a sizeable page on its own, but given the current structure and pages, it seems a bit out of place. 🙈 
But in any case, given this PR context, I don't think we should be changing anything about it, WDYT?

Reply:
Not in this PR clearly yes :+1:

Reply:
Done